lib/improver/tests/argparser/test_ArgParser.py

```diff
@@ -0,0 +1,328 @@
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for argparser.ArgParser."""
+
+
+import os
+import unittest
+from mock import patch
+
+from improver.argparser import ArgParser
+
+
+# We might one day want to move this up to a more central place.
+class QuietTestCase(unittest.TestCase):
+
+    """A subclass of unittest.TestCase which prevents writing to stderr and
+    calling of sys.exit."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Patch the class by redirecting stderr to /dev/null, and disabling
+        calls to sys.exit."""
+        cls.stderr_patch = patch('sys.stderr', open(os.devnull, 'w'))
+        cls.exit_patch = patch('sys.exit')
+        cls.stderr_patch.start()
+        cls.exit_patch.start()
+
+    @classmethod
+    def tearDownClass(cls):
+        """Stop the patches which redirect stderr to /dev/null and prevents
+        sys.exit from being called."""
+        cls.stderr_patch.stop()
+        cls.exit_patch.stop()
+
+
+class Test_init(QuietTestCase):
+
+    """Test the __init__ method."""
+
+    def test_create_argparser_with_no_arguments(self):
+        """Test that creating an ArgParser with no arguments has no
+        arguments."""
+
+        compulsory_arguments = {}
+
+        # it doesn't matter what the centralized arguments are, because we
+        # select None of them
+        with patch('improver.argparser.ArgParser.COMPULSORY_ARGUMENTS',
+                   compulsory_arguments):
+            parser = ArgParser(central_arguments=None, specific_arguments=None)
+            args = parser.parse_args()
+            args = vars(args).keys()
+            self.assertEqual(len(args), 0)
+
+    def test_create_argparser_only_compulsory_arguments(self):
+        """Test that creating an ArgParser with only compulsory arguments
+        adds only the compulsory arguments."""
+
+        compulsory_arguments = {'foo': (['--foo'], {})}
+
+        # it doesn't matter what the centralized arguments are, because we
+        # select None of them
+        with patch('improver.argparser.ArgParser.COMPULSORY_ARGUMENTS',
+                   compulsory_arguments):
+            parser = ArgParser(central_arguments=None, specific_arguments=None)
+            args = parser.parse_args()
+            args = vars(args).keys()
+            self.assertItemsEqual(args, ['foo'])
+
+    def test_create_argparser_fails_with_unknown_centralized_argument(self):
+        """Test that we raise an exception when attempting to retrieve
+        centralized arguments which are not centralized argument dictionary."""
+
+        centralized_arguments = {'foo': (['--foo'], {})}
+
+        central_args_to_fetch = ('missing_central_arg',)
+        with patch('improver.argparser.ArgParser.CENTRALIZED_ARGUMENTS',
+                   centralized_arguments):
```

Comment:
I'd argue we don't need to mock the class variable here - it's highly unlikely we'd ever include a 'missing_central_arg' in the centralised list.  And it would be good to cover the class variables as well as method behaviour in these unit tests.